### PR TITLE
feat: Transition export workflows to use torch._export APIs 

### PR DESCRIPTION
**Pull request by peri044**
# Description

* Use torch._export APIs to trace instead of dynamo export. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/__init__.py [hunk 1]
```diff
@@ -3,10 +3,12 @@
...
 
 if version.parse(sanitized_torch_version()) >= version.parse("2.1.dev"):
-    from ._settings import *  # noqa: F403
-    from ._SourceIR import SourceIR  # noqa: F403
-    from .aten_tracer import trace  # noqa: F403
-    from .compile import compile  # noqa: F403
-    from .conversion import *  # noqa: F403
-    from .conversion.converter_registry import DYNAMO_CONVERTERS  # noqa: F403
-    from .conversion.converter_registry import dynamo_tensorrt_converter  # noqa: F403
+    from ._settings import *
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/__init__.py:6]: You need the noqa to pass ruff

File: py/torch_tensorrt/dynamo/aten_tracer.py [hunk 2]
```diff
@@ -1,157 +1,27 @@
...
-        pr: PassResult = passes(fx_module)
-        fx_module = pr.graph_module
-
-    fx_module(*inputs)
-
-    fx_module = run_const_fold(fx_module)
-    print(fx_module.graph)
-    return fx_module
+    fake_mode = FakeTensorMode(allow_non_fake_inputs=True)
+    with unittest.mock.patch.object(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/aten_tracer.py:18]: Why do we use unittest here?

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/aten_tracer.py:18]: This was how inductor used to enable `allow_non_fake_inputs` from within a `torch.compile` execution [as here](https://github.com/pytorch/pytorch/blob/750b9b359f06cb8b8c2d5b6118bba636e2112cbb/torch/_inductor/freezing.py#L176-L178), but this code has since been removed from that file.

This paradigm does appear to be used in other places in the Dynamo codebase, such as [here](https://github.com/pytorch/pytorch/blob/6af6b8f728426fb7551630e28148c0017fa501bc/torch/_guards.py#L523-L525).

There was an ongoing discussion https://github.com/pytorch/pytorch/issues/105485#issuecomment-1642844979 regarding the topic of fake tensor modes.

Comment (gs-olive) [inline py/torch_tensorrt/dynamo/aten_tracer.py:18]: Consider adding a gradient context here, to mimic the behavior of `aot_export_joint_simple` as in [this snippet](https://github.com/pytorch/pytorch/blob/6af6b8f728426fb7551630e28148c0017fa501bc/torch/_functorch/aot_autograd.py#L4016-L4027)

File: py/torch_tensorrt/dynamo/aten_tracer.py [hunk 3]
```diff
@@ -1,157 +1,27 @@
...
-                "The user code is using a feature we don't support. "
-                "Please try torchdynamo.explain() to get possible the reasons",
-            ) from exc
-        except Exception as exc:
-            raise RuntimeError(
-                "torchdynamo internal error occured. Please see above stacktrace"
-            ) from exc
-
-
-@req_torch_version("2.dev")
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/aten_tracer.py:124]: Should this required version decorator be changed to a more recent Torch version?

File: py/torch_tensorrt/dynamo/aten_tracer.py [hunk 4]
```diff
@@ -1,160 +1,32 @@
...
-        fx_module = pr.graph_module
-
-    fx_module(*inputs)
-
-    fx_module = run_const_fold(fx_module)
-    logger.info("Post export graph : %s\n", fx_module.graph)
-    return fx_module
+    # Set log level at the top of compilation (torch_tensorrt.dynamo)
+    if ("debug" in kwargs and kwargs["debug"]) and logger.parent:
+        logger.parent.setLevel(logging.DEBUG)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/aten_tracer.py:22]: Is this necessary to set in the `aten_tracer`?

File: tests/py/dynamo/models/test_models_export.py [hunk 5]
```diff
@@ -118,6 +114,59 @@ def test_efficientnet_b0(ir):
...
+        assertions.assertTrue(
+            cos_sim > COSINE_THRESHOLD,
+            msg=f"HF BERT base-uncased TRT outputs don't match with the original model. Cosine sim score: {cos_sim} Threshold: {COSINE_THRESHOLD}",
+        )
+
+    # Clean up model env
+    torch._dynamo.reset()
+
+    with torch.no_grad():
+        torch.cuda.empty_cache()
```
Comment (gs-olive) [inline tests/py/dynamo/models/test_models_export.py:167]: This can be removed (also from the other test cases). I do not think it is recommended to use the `empty_cache` function anymore.

File: tests/py/dynamo/models/test_models_export.py [hunk 6]
```diff
@@ -118,6 +114,59 @@ def test_efficientnet_b0(ir):
...
+                dtype=input.dtype,
+                format=torch.contiguous_format,
+            ),
+        ],
+        "device": torchtrt.Device("cuda:0"),
+        "enabled_precisions": {torch.float},
+        "truncate_long_and_double": True,
+        "ir": ir,
+        "min_block_size": 10,
+        "torch_executed_ops": {"torch.ops.aten.gelu.default"},
```
Comment (gs-olive) [inline tests/py/dynamo/models/test_models_export.py:146]: This line can be removed, since the decompositions should handle this

**Review by gs-olive — COMMENTED** [review]
Looks good! Left some comments and general suggestions.

File: py/torch_tensorrt/dynamo/aten_tracer.py (hunk 4 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/aten_tracer.py:22]: Yes. this is because aten_tracer.trace is the entry point for ir=dynamo flow. I'll try to make this as a function as we are using this in 3 places now.

File: tests/py/dynamo/models/test_models_export.py (hunk 5 above)
Comment (peri044) [reply tests/py/dynamo/models/test_models_export.py:167]: done

File: tests/py/dynamo/models/test_models_export.py (hunk 6 above)
Comment (peri044) [reply tests/py/dynamo/models/test_models_export.py:146]: got it.

**Review by gs-olive — APPROVED** [review]
Looks good to me!

**Comment by gs-olive** [comment]
For some reason this error is not being shown on `main`, but I think you could change:
```python
import torch_tensorrt

... torch_tensorrt.sanitized_torch_version() ...
```
To:
```python
from torch_tensorrt._utils import sanitized_torch_version

... sanitized_torch_version() ...
```
